x/marketmaker/keeper/marketmaker.go

```diff
@@ -0,0 +1,356 @@
+		k.SetMarketMaker(ctx, types.MarketMaker{
+			Address:  mmAddr.String(),
+			PairId:   pairId,
+			Eligible: false,
+		})
+	}
+
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.EventTypeApplyMarketMaker,
+			sdk.NewAttribute(types.AttributeKeyAddress, mmAddr.String()),
+			sdk.NewAttribute(types.AttributeKeyPairIds, strings.Trim(strings.Replace(fmt.Sprint(pairIds), " ", ",", -1), "[]")),
+		),
+		sdk.NewEvent(
+			sdk.EventTypeMessage,
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.AttributeValueCategory),
+			sdk.NewAttribute(sdk.AttributeKeyAction, types.EventTypeApplyMarketMaker),
+			sdk.NewAttribute(sdk.AttributeKeySender, mmAddr.String()),
+		),
+	})
+
+	return nil
+}
+
+func (k Keeper) ClaimIncentives(ctx sdk.Context, mmAddr sdk.AccAddress) error {
+	incentive, found := k.GetIncentive(ctx, mmAddr)
+	if !found {
+		return types.ErrEmptyClaimableIncentive
+	}
+
+	if err := k.bankKeeper.SendCoins(ctx, types.ClaimableIncentiveReserveAcc, mmAddr, incentive.Claimable); err != nil {
+		return err
+	}
+
+	k.DeleteIncentive(ctx, mmAddr)
+
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.EventTypeClaimIncentives,
+			sdk.NewAttribute(types.AttributeKeyAddress, mmAddr.String()),
+			sdk.NewAttribute(sdk.AttributeKeyAmount, incentive.Claimable.String()),
+		),
+		sdk.NewEvent(
+			sdk.EventTypeMessage,
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.AttributeValueCategory),
+			sdk.NewAttribute(sdk.AttributeKeyAction, types.EventTypeClaimIncentives),
+			sdk.NewAttribute(sdk.AttributeKeySender, mmAddr.String()),
+		),
+	})
+	return nil
+}
+
+func (k Keeper) ValidateDepositReservedAmount(ctx sdk.Context) error {
+	mmCount := 0
+	depositCount := 0
+	var totalAmt sdk.Coins
+	k.IterateMarketMakers(ctx, func(mm types.MarketMaker) (stop bool) {
+		if !mm.Eligible {
+			mmCount += 1
+		}
+		return false
+	})
+	k.IterateDeposits(ctx, func(id types.Deposit) (stop bool) {
+		depositCount += 1
+		totalAmt = totalAmt.Add(id.Amount...)
+		return false
+	})
+	if mmCount != depositCount {
+		return fmt.Errorf("market maker number differs from the actual value; have %d, want %d", mmCount, depositCount)
+	}
+
+	reserveBalance := k.bankKeeper.GetAllBalances(ctx, types.DepositReserveAcc)
+	if !totalAmt.Empty() && reserveBalance.IsAllLT(totalAmt) {
+		return fmt.Errorf("DepositReserveAcc differs from the actual value; have %s, want %s", reserveBalance, totalAmt)
+	}
+	return nil
+}
+
+func (k Keeper) ValidateIncentiveReservedAmount(ctx sdk.Context, incentives []types.Incentive) error {
+	var totalClaimable sdk.Coins
+	for _, record := range incentives {
+		totalClaimable = totalClaimable.Add(record.Claimable...)
+	}
+
+	reserveBalance := k.bankKeeper.GetAllBalances(ctx, types.ClaimableIncentiveReserveAcc)
+	if !totalClaimable.Empty() && reserveBalance.IsAllLT(totalClaimable) {
+		return fmt.Errorf("ClaimableIncentiveReserveAcc differs from the actual value; have %s, want %s", reserveBalance, totalClaimable)
+	}
```

Comment:
I think we can avoid using `GetAllBalances` if `totalClaimable` is empty. Same suggestion as above comment.